lib/taste_tester/commands.rb

```diff
@@ -99,80 +97,88 @@ def self.test
       end
       unless TasteTester::Config.skip_pre_test_hook ||
           TasteTester::Config.linkonly
-        TasteTester::Hooks.pre_test(TasteTester::Config.dryrun, repo, hosts)
-      end
-      tested_hosts = []
-      hosts.each do |hostname|
-        host = TasteTester::Host.new(hostname, server)
-        begin
-          host.test
-          tested_hosts << hostname
-        rescue TasteTester::Exceptions::AlreadyTestingError => e
-          logger.error("User #{e.username} is already testing on #{hostname}")
-        end
+        TasteTester::Hooks.pre_test(TasteTester::Config.dryrun, repo,
+                                    TasteTester::Config.servers)
       end
+      hosts = _run_on_hosts(:test)
       unless TasteTester::Config.skip_post_test_hook ||
           TasteTester::Config.linkonly
         TasteTester::Hooks.post_test(TasteTester::Config.dryrun, repo,
-                                     tested_hosts)
+                                     hosts.select { |_, r| r == :ok }.keys)
       end
-      # Strictly: hosts and tested_hosts should be sets to eliminate variance in
-      # order or duplicates. The exact comparison works here because we're
-      # building tested_hosts from hosts directly.
-      if tested_hosts == hosts
+      if hosts.values.all? :ok
         # No exceptions, complete success: every host listed is now configured
         # to use our chef-zero instance.
         exit(0)
       end
-      if tested_hosts.empty?
+      if hosts.values.none? :ok
+        if hosts.values.any? :ssh_error
+          # All the hosts we had failed, with at least one because of ssh
+          exit(1)
+        end
         # All requested hosts are being tested by another user. We didn't change
         # their configuration.
         exit(3)
       end
       # Otherwise, we got a mix of success and failure due to being tested by
-      # another user. We'll be pessemistic and return an error because the
+      # another user. We'll be pessimistic and return an error because the
       # intent to taste test the complete list was not successful.
-      # code.
       exit(2)
     end
 
-    def self.untest
+    def self._run_on_hosts(method)
       hosts = TasteTester::Config.servers
       unless hosts
         logger.error('You must provide a hostname')
         exit(1)
       end
       server = TasteTester::Server.new
-      hosts.each do |hostname|
-        host = TasteTester::Host.new(hostname, server)
-        host.untest
+      hosts_result = {}
+      host_threads = []
+      queue = Queue.new
+      hosts.each { |hostname| queue << hostname }
+      [TasteTester::Config.parallel_hosts, hosts.length].min.times do
+        host_threads << Thread.new do
+          Thread.current.report_on_exception = false
+          loop do
+            begin
+              hostname = queue.pop(true)
```

Comment:
do we need like `break unless hostname` here?

Reply:
We fill the queue with hostnames 3 lines above, and if it's empty it will throw and we rescue it.
I don't see how this might be empty, but on the other hand, it doesn't really hurt either :)

Reply:
what happens when the queue is empty though? And you're out of work?